src/main/java/org/hmhb/authentication/DefaultAuthenticationService.java

```diff
@@ -0,0 +1,103 @@
+
+import static java.util.Objects.requireNonNull;
+
+@Service
+public class DefaultAuthenticationService implements AuthenticationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultAuthenticationService.class);
+
+    private final Environment environment;
+    private final GoogleOauthService googleOauthService;
+    private final JwtAuthenticationService jwtAuthenticationService;
+    private final UserService userService;
+
+    @Autowired
+    public DefaultAuthenticationService(
+            @Nonnull Environment environment,
+            @Nonnull GoogleOauthService googleOauthService,
+            @Nonnull JwtAuthenticationService jwtAuthenticationService,
+            @Nonnull UserService userService
+    ) {
+        this.environment = requireNonNull(environment, "environment cannot be null");
+        this.googleOauthService = requireNonNull(googleOauthService, "googleOauthService cannot be null");
+        this.jwtAuthenticationService = requireNonNull(jwtAuthenticationService, "jwtAuthenticationService cannot be null");
+        this.userService = requireNonNull(userService, "userService cannot be null");
+    }
+
+    @Override
+    public TokenResponse authenticate(
+            @Nonnull GoogleOauthAccessRequestInfo request
+    ) {
+        LOGGER.debug("authenticate called: request={}", request);
+        requireNonNull(request, "request cannot be null");
+
+        String email;
+
+        String clientId = environment.getProperty("google.oauth.client.id");
+        String clientSecret = environment.getProperty("google.oauth.client.secret");
+
+        if (!clientId.equals(request.getClientId())) {
+            throw new ClientIdMismatchException(
+                    request.getClientId(),
+                    clientId
+            );
+        }
+
+        LOGGER.debug(
+                "Calling into google: clientId={}, code={}, redirectUrl={}",
+                request.getClientId(),
+                request.getCode(),
+                request.getRedirectUri()
+        );
+
+        GoogleTokenResponse googleTokenResponse = googleOauthService.getTokenResponse(
+                request.getClientId(),
+                clientSecret,
+                request.getCode(),
+                request.getRedirectUri()
+        );
+
+        try {
+            GoogleIdToken idToken = googleTokenResponse.parseIdToken();
+            email = idToken.getPayload().getEmail();
+            LOGGER.debug("successfully called into google: email={}", email);
+        } catch (IOException e) {
+            throw new GoogleOauthException("Failed to parse google id token!", e);
+        }
+
+        HmhbUser user;
+
+        try {
+            user = userService.getUserByEmail(email);
+        } catch (UserNotFoundException e) {
+            user = userService.provisionNewUser(email);
+        }
+
+        TokenResponse tokenResponse = new TokenResponse(
+                jwtAuthenticationService.generateJwtToken(user)
```

Comment:
We encode the user information into a jwt token so we can later use that information w/o having to ask google again.